tsdb/engine/wal/wal.go

```diff
@@ -0,0 +1,1591 @@
+/*
+Package WAL implements a write ahead log optimized for write throughput
+that can be put in front of the database index.
+
+The WAL is broken into different partitions. The default number of
+partitions is 5. Each partition consists of a number of segment files.
+By default these files will get up to 2MB in size before a new segment
+file is opened. The files are numbered and start at 1. The number
+indicates the order in which the files should be read on startup to
+ensure data is recovered in the same order it was written.
+
+Partitions are flushed and compacted individually. One of the goals with
+having multiple partitions was to be able to flush only a portion of the
+WAL at a time.
+
+The WAL does not flush everything in a partition when it comes time. It will
+only flush series that are over a given threshold (32kb by default). The rest
+will be written into a new segment file so they can be flushed later. This
+is like a compaction in an LSM Tree.
+*/
+package wal
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"hash/fnv"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/golang/snappy"
+	"github.com/influxdb/influxdb/tsdb"
+)
+
+const (
+	// DefaultSegmentSize of 2MB is the size at which segment files will be rolled over
+	DefaultSegmentSize = 2 * 1024 * 1024
+
+	// PartitionCount is the number of partitions in the WAL
+	PartitionCount = 5
```

Comment:
Just a number I pulled out. I did some rough testing with different numbers. If it's too many then we end up with too many open files. If it's too low then we end up flushing and compacting way too much of the keyspace in one go.

It would be worth doing some more testing to see if there's a better number, but this is good enough for now.